lib/streamlit/caching.py

```diff
@@ -80,6 +77,57 @@ def __init__(self, cached_value):
 DiskCacheEntry = namedtuple("DiskCacheEntry", ["value"])
 
 
+class _MemCaches(object):
+    """Manages all in-memory st.cache caches"""
+
+    def __init__(self):
+        # Contains a cache object for each st.cache'd function
+        self._lock = threading.RLock()
+        self._function_caches = {}  # type: Dict[Any, TTLCache]
+
+    def get_cache(self, key, max_entries, ttl) -> TTLCache:
+        """Return the mem cache for the given key.
+
+        If it doesn't exist, create a new one with the given params.
+        """
+
+        if max_entries is None:
+            max_entries = math.inf
+        if ttl is None:
+            ttl = math.inf
+
+        # Get the existing cache, if it exists, and validate that its params
+        # haven't changed.
+        with self._lock:
+            mem_cache = self._function_caches.get(key)
+            if (
+                mem_cache is not None
+                and mem_cache.ttl == ttl
+                and mem_cache.maxsize == max_entries
+            ):
+                return mem_cache
+
+            # Create a new cache object and put it in our dict
+            LOGGER.debug(
+                "Creating new mem_cache (key=%s, max_entries=%s, ttl=%s)",
+                key,
+                max_entries,
+                ttl,
+            )
+            mem_cache = TTLCache(maxsize=max_entries, ttl=ttl, timer=TTLCACHE_TIMER)
```

Comment:
The timer types are just functions that return values. (So: time.time(), time.monotonic(), etc.) The only reason it's explicitly specified here is so that the passed-in timer can be patched in unit tests. I'll add a comment to that effect!

Reply:
> Changing any of the cache params trashes all the previous cache entries? Is this communicated to the user somewhere?

Not explicitly, no. Though changing anything else about the function (its name, its contents) also trashes the cache, so this seems like a reasonable thing to do without spelling it out. I'm willing to change it if there's disagreement, though!